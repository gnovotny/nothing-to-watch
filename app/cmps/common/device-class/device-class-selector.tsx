import { TabletSmartphone } from 'lucide-react'
import { useMemo } from 'react'
import { cn } from '../../../utils/tw'
import { DEVICE_CLASS, DEVICE_CLASS_ITEMS } from '../../../vf/consts'
import { Badge } from '../../ui/badge'
import { Selector, type SelectorItems } from '../selector'

export function DeviceClassSelector({
  className = '',
  value,
  estimatedValue,
  onValueChange,
}: {
  className?: string
  value?: DEVICE_CLASS
  estimatedValue?: DEVICE_CLASS
  onValueChange: (value: DEVICE_CLASS) => void
}) {
  const deviceClassItems: SelectorItems = useMemo(() => {
    return DEVICE_CLASS_ITEMS.filter((deviceClass) => {
      if (estimatedValue === DEVICE_CLASS.mobile) {
        return deviceClass.id < DEVICE_CLASS.high
      }
      return deviceClass.id !== DEVICE_CLASS.mobile
    }).map((deviceClass) => {
      return {
        label: deviceClass.name,
        value: String(deviceClass.id),
        addon:
          estimatedValue === deviceClass.id ? (
            <Badge className='-translate-x-1/2 -translate-y-1/2 !text-background pointer-events-none absolute top-0 left-1/2 text-xxs'>
              Estimated
            </Badge>
          ) : null,
      }
    })
  }, [estimatedValue])

  return (
    <div className='flex flex-col gap-4'>
      <div className='flex items-center gap-2 font-semibold text-xl text-zinc-900 dark:text-white'>
        <TabletSmartphone className='h-5 w-5 text-zinc-900 dark:text-white' />
        Your device
      </div>
      <Selector
        className={cn('', className)}
        defaultValue={String(value)}
        onValueChange={(value) => {
          onValueChange(Number.parseInt(value) as DEVICE_CLASS)
        }}
        items={deviceClassItems}
      />
    </div>
  )
}
