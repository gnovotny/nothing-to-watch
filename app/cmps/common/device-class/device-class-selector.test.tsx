import { render, screen } from '@testing-library/react'
import type React from 'react'
import { describe, expect, it, vi } from 'vitest'
import { DeviceClassSelector } from './device-class-selector'

// Mock the Selector component
vi.mock('../selector', () => ({
  Selector: ({
    className,
    defaultValue,
    onValueChange,
    items,
    ...props
  }: {
    className?: string
    defaultValue?: string
    onValueChange?: (value: string) => void
    items?: Array<{ value: string; label: string; addon?: React.ReactNode }>
    [key: string]: unknown
  }) => (
    <div
      className={className}
      data-testid='device-class-selector'
      data-value={defaultValue}
      onClick={() => onValueChange?.(defaultValue || '')}
      {...props}
    >
      {items?.map(
        (
          item: { value: string; label: string; addon?: React.ReactNode },
          index: number,
        ) => (
          <div key={item.value || index} data-value={item.value}>
            {item.label}
            {item.addon}
          </div>
        ),
      )}
    </div>
  ),
}))

// Mock device class constants and utilities
vi.mock('../../../vf/consts', () => ({
  DEVICE_CLASS: {
    1: 1,
    2: 2,
    3: 3,
    4: 4,
  },
  DEVICE_CLASS_LABELS: {
    1: 'Low-end',
    2: 'Mid-range',
    3: 'High-end',
    4: 'Flagship',
  },
  DEVICE_CLASS_ITEMS: [
    { value: 1, label: 'Low-end' },
    { value: 2, label: 'Mid-range' },
    { value: 3, label: 'High-end' },
    { value: 4, label: 'Flagship' },
  ],
}))

vi.mock('../../../vf/utils', () => ({
  getDeviceClassLabel: vi.fn((deviceClass: number) => {
    const labels = {
      1: 'Low-end',
      2: 'Mid-range',
      3: 'High-end',
      4: 'Flagship',
    }
    return labels[deviceClass as keyof typeof labels] || 'Unknown'
  }),
}))

describe('DeviceClassSelector', () => {
  const mockOnChange = vi.fn()

  beforeEach(() => {
    mockOnChange.mockClear()
  })

  it('should render with default props', () => {
    render(<DeviceClassSelector value={2} onValueChange={mockOnChange} />)

    expect(screen.getByText('Your device')).toBeInTheDocument()
    expect(screen.getByTestId('device-class-selector')).toBeInTheDocument()
  })

  it('should render with custom className', () => {
    render(
      <DeviceClassSelector
        value={3}
        onValueChange={mockOnChange}
        className='custom-class'
      />,
    )

    const selector = screen.getByTestId('device-class-selector')
    expect(selector).toHaveClass('custom-class')
  })

  it('should have correct default value', () => {
    render(<DeviceClassSelector value={3} onValueChange={mockOnChange} />)

    const selector = screen.getByTestId('device-class-selector')
    expect(selector).toHaveAttribute('data-value', '3')
  })

  it('should match snapshot with deterministic props', () => {
    const { container } = render(
      <DeviceClassSelector
        value={2}
        onValueChange={mockOnChange}
        className='test-class'
      />,
    )

    expect(container.firstChild).toMatchSnapshot()
  })

  it('should match snapshot with different device class values', () => {
    const deviceClasses = [1, 2, 3, 4]

    deviceClasses.forEach((deviceClass) => {
      const { container } = render(
        <DeviceClassSelector
          value={deviceClass}
          onValueChange={mockOnChange}
        />,
      )

      expect(container.firstChild).toMatchSnapshot(
        `device-class-${deviceClass}`,
      )
    })
  })
})
