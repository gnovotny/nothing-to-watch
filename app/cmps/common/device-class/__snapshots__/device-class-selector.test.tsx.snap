// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`DeviceClassSelector > should match snapshot with deterministic props 1`] = `
<div
  class="flex flex-col gap-4"
>
  <div
    class="flex items-center gap-2 font-semibold text-xl text-zinc-900 dark:text-white"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-tablet-smartphone h-5 w-5 text-zinc-900 dark:text-white"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="14"
        rx="2"
        width="10"
        x="3"
        y="8"
      />
      <path
        d="M5 4a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v16a2 2 0 0 1-2 2h-2.4"
      />
      <path
        d="M8 18h.01"
      />
    </svg>
    Your device
  </div>
  <div
    class="test-class"
    data-testid="device-class-selector"
    data-value="2"
  />
</div>
`;

exports[`DeviceClassSelector > should match snapshot with different device class values > device-class-1 1`] = `
<div
  class="flex flex-col gap-4"
>
  <div
    class="flex items-center gap-2 font-semibold text-xl text-zinc-900 dark:text-white"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-tablet-smartphone h-5 w-5 text-zinc-900 dark:text-white"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="14"
        rx="2"
        width="10"
        x="3"
        y="8"
      />
      <path
        d="M5 4a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v16a2 2 0 0 1-2 2h-2.4"
      />
      <path
        d="M8 18h.01"
      />
    </svg>
    Your device
  </div>
  <div
    class=""
    data-testid="device-class-selector"
    data-value="1"
  />
</div>
`;

exports[`DeviceClassSelector > should match snapshot with different device class values > device-class-2 1`] = `
<div
  class="flex flex-col gap-4"
>
  <div
    class="flex items-center gap-2 font-semibold text-xl text-zinc-900 dark:text-white"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-tablet-smartphone h-5 w-5 text-zinc-900 dark:text-white"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="14"
        rx="2"
        width="10"
        x="3"
        y="8"
      />
      <path
        d="M5 4a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v16a2 2 0 0 1-2 2h-2.4"
      />
      <path
        d="M8 18h.01"
      />
    </svg>
    Your device
  </div>
  <div
    class=""
    data-testid="device-class-selector"
    data-value="2"
  />
</div>
`;

exports[`DeviceClassSelector > should match snapshot with different device class values > device-class-3 1`] = `
<div
  class="flex flex-col gap-4"
>
  <div
    class="flex items-center gap-2 font-semibold text-xl text-zinc-900 dark:text-white"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-tablet-smartphone h-5 w-5 text-zinc-900 dark:text-white"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="14"
        rx="2"
        width="10"
        x="3"
        y="8"
      />
      <path
        d="M5 4a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v16a2 2 0 0 1-2 2h-2.4"
      />
      <path
        d="M8 18h.01"
      />
    </svg>
    Your device
  </div>
  <div
    class=""
    data-testid="device-class-selector"
    data-value="3"
  />
</div>
`;

exports[`DeviceClassSelector > should match snapshot with different device class values > device-class-4 1`] = `
<div
  class="flex flex-col gap-4"
>
  <div
    class="flex items-center gap-2 font-semibold text-xl text-zinc-900 dark:text-white"
  >
    <svg
      aria-hidden="true"
      class="lucide lucide-tablet-smartphone h-5 w-5 text-zinc-900 dark:text-white"
      fill="none"
      height="24"
      stroke="currentColor"
      stroke-linecap="round"
      stroke-linejoin="round"
      stroke-width="2"
      viewBox="0 0 24 24"
      width="24"
      xmlns="http://www.w3.org/2000/svg"
    >
      <rect
        height="14"
        rx="2"
        width="10"
        x="3"
        y="8"
      />
      <path
        d="M5 4a2 2 0 0 1 2-2h12a2 2 0 0 1 2 2v16a2 2 0 0 1-2 2h-2.4"
      />
      <path
        d="M8 18h.01"
      />
    </svg>
    Your device
  </div>
  <div
    class=""
    data-testid="device-class-selector"
    data-value="4"
  />
</div>
`;
